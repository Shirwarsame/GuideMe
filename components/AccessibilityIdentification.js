import React from 'react';
import { Text, View, Image } from 'react-native';
import { Marker } from 'react-native-maps';
import { styles } from '../screens/Map'


function AccessibilityIdentification() {
    let handicapedSignSource = './../assets/Icons/handicap_icon.png';

        return(
            <View>
                    <Marker
                    coordinate={{ latitude: 45.497245, longitude: -73.578968}}>
                    <Image style={styles.AccessibilityIdentification} source={require(handicapedSignSource)}/>
                    </Marker>

                    <Marker
                    coordinate={{ latitude: 45.496612, longitude: -73.578010}}>
                    <Image style={styles.AccessibilityIdentification} source={require(handicapedSignSource)}/>
                    </Marker>

                    <Marker
                    coordinate={{ latitude: 45.495899, longitude:  -73.578712}}>
                    <Image style={styles.AccessibilityIdentification} source={require(handicapedSignSource)}/>
                    </Marker>
            </View>
            );
    }


export {AccessibilityIdentification};