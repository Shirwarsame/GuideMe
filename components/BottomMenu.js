/* eslint-disable no-unused-vars */
import React, { useLayoutEffect, useEffect } from "react";
import { View, AsyncStorage, Text, StyleSheet, Switch } from "react-native";
import { Icon } from "native-base";
import { Button } from "react-native-paper";
import { store } from "../redux/reducers/index";
import { FloorMenu } from "./FloorMenu";
import {CampusRegion} from "../constants/buildingData";


/**
 * US6 - As a user, I would like to switch between the SGW and the Loyola maps
 * The following function renders a menu at the bottom of the screen. The menu
 * includes a toggle (US6) & an arrow icon leading to the More Details page.
 */

function BottomMenu (props) {
    const [selectedBuilding, setSelectedBuilding] = React.useState(null);
    const [swtichCampus, setSwitchCampus] = React.useState(true);
    const [destination, setDestination] = React.useState("");
    const [methodTravel, setMethodTravel] = React.useState("driving");
    const [personaType, setPersonaType] = React.useState("");
    const [mobilityReduced, setMobilityReduced] = React.useState("");
    const viewIndoor = props.indoor;
    const inDirections = props.inDirections;
    const building = props.building;
    const previewDirections = props.previewMode;
    const from = props.from;
    const to = props.to;
    

    const goBack = () => {
        props.navigation.goBack();
    };

    const goToDoubleSearchBar = () => {
        props.navigation.navigate("DoubleSearch", { destinationName: destination });
    };
    const goToDirections = () => {
        props.navigation.navigate("Directions", { destinationResponse: props.directionResponse });
    };

    const goToPreferenceMenu = (isIndoor) => {
        if (isIndoor) {
            props.navigation.navigate("PreferenceMenu", {
                indoor: true,
                personaType: personaType,
                mobilityType: mobilityReduced,
                transportType: methodTravel
            });
        } else {
            props.navigation.navigate("PreferenceMenu", {
                indoor: false,
                personaType: personaType,
                mobilityType: mobilityReduced,
                transportType: methodTravel
            });
        }

    };

    const goToMoreDetails = () => {
        props.navigation.navigate("MoreDetails", {
            name: store.getState().selectedBuildingName
        });
    };

    const goToInsideBuilding = () => {
        props.navigation.navigate("IndoorMapView", { selectedBuilding: selectedBuilding });
    };

    const goToNearby = () => {
        AsyncStorage.setItem("sideMenu", "mapView"); 
        props.navigation.navigate("NearbyInterest", { campusBool: swtichCampus});
    
    };

    
    useLayoutEffect(() => {
        const unsubscribe = store.subscribe(() => {
            setSelectedBuilding(store.getState().selectedBuildingName);
            setDestination(store.getState().mainSearchBarDestination);
            setPersonaType(store.getState().personaType);
            setMobilityReduced(store.getState().mobilityReducedType);
            setMethodTravel(store.getState().transportType);
        });
        return function cleanUp() {
            unsubscribe();
        }
    });


    const switchCampuses = (value) => {
        setSwitchCampus(value);
        if (value) {
            props.mapReference.current.animateToRegion(CampusRegion.sgwCoord);        }
        else {
            props.mapReference.current.animateToRegion(CampusRegion.loyCoord);
        }
    }

    if (viewIndoor) {
        return (
            <View style={styles.insideBuildingContainer} data-test="BottomMenu" testID="bottomMenuInitalView">
                <Icon name="ios-arrow-up" style={styles.arrowUp} onPress={() => { (inDirections ? goToPreferenceMenu(true) : goToMoreDetails) }} />
                <Text style={styles.mainLabel}>{building ? building : selectedBuilding}</Text>
                <Text style={styles.shortLabel}>More info</Text>
                <Button testID="indoorMapExitBuildingButton" style={styles.btnleave} color={"#3ACCE1"} uppercase={false} mode="contained" onPress={goBack}>
                    <Text style={styles.btnText}>Exit Building</Text>
                </Button>
                <View style={styles.changeFloor}>
                    <FloorMenu from={from} to={to} />
                </View>
            </View>
        )
    }

    if (previewDirections) {
        return (
            <View style={styles.container}>
                <Icon name="ios-arrow-up" style={styles.arrowUp} onPress={() => { goToPreferenceMenu(false) }} />

                {props.directionResponse && props.directionResponse.generalRouteInfo.totalDuration.length > 12 &&
                    <Text style={styles.mainLabel}>{props.directionResponse.generalRouteInfo.totalDuration}</Text>
                }
                 {props.directionResponse && props.directionResponse.generalRouteInfo.totalDuration.length <= 12 &&
                    <Text style={styles.mainLabel}>{props.directionResponse.generalRouteInfo.totalDuration } {props.directionResponse.generalRouteInfo.totalDistance}</Text>
                }
                 {!props.directionResponse &&
                    <Text style={styles.mainLabel}>N/A</Text>
                }
                <Text style={styles.shortLabel}>Main Travel Mode: {methodTravel}</Text>
                <View style={styles.btnGetDirection}>
                    <Button style={styles.btnGetDirectionPosition}
                        color={"#3ACCE1"} uppercase={false} mode="contained" onPress={goToDirections}>
                        <Text style={styles.btnText}>Start</Text>
                    </Button>
                </View>
            </View>
        );
    }
    if (destination) {
        if (destination.length > 13) {
            var updatedDestination = destination.substring(0, 13) + "...";
        }
        return (
            <View style={styles.container} >
                <Icon name="ios-arrow-up" style={styles.arrowUp} onPress={() => { props.navigation.navigate("MoreDetails", { name: destination }) }} />
                {destination.length > 13
                    ? <Text style={styles.mainLabel}>{updatedDestination}</Text>
                    : <Text style={styles.mainLabel}>{destination}</Text>
                }
                <Text style={styles.shortLabel}>More info</Text>
                <View style={styles.btnGetDirection}>
                    <Button testID="getDirectionsButton" style={styles.btnGetDirection} color={"#3ACCE1"} uppercase={false} mode="contained" onPress={goToDoubleSearchBar}>
                        <Text style={styles.btnText}>Get Directions</Text>
                    </Button>
                </View>
            </View>
        );
    }

    if (selectedBuilding) {
        if (selectedBuilding.length > 13) {
            var updatedSelectedBuilding = selectedBuilding.substring(0, 13) + "...";

        }
        return (
            <View style={styles.container}>
                <Icon testID="getInsideIcon" name="ios-arrow-up" style={styles.arrowUp} onPress={goToMoreDetails} />
                <Text style={styles.mainLabel}>{selectedBuilding}</Text>
                {selectedBuilding.length > 13
                    ? <Text style={styles.mainLabel}>{updatedSelectedBuilding}</Text>
                    : <Text style={styles.mainLabel}>{selectedBuilding}</Text>
                }
                <Text style={styles.shortLabel}>More info</Text>
                <Button testID="getInsideButton" style={styles.btn} color={"#3ACCE1"} uppercase={false} mode="contained" onPress={goToInsideBuilding}>
                    <Text style={styles.btnText}>Get Inside</Text>
                </Button>
            </View>
        );
    }
    else {
        return (
            <View style={styles.container} data-test="BottomMenu" testID="bottomMenuInitalView">
                <Icon name="ios-arrow-up" style={styles.arrowUp} onPress={goToNearby} />
                <Text style={styles.mainLabel}>Nearby</Text>
                <Text style={styles.shortLabel}>Food, drinks & more</Text>
                <View testID="intialBottomMenuToggleButton" style={styles.toggle}>
                    <Switch
                        value={swtichCampus}
                        onValueChange={(val) => { switchCampuses(val);}}>
                    </Switch>
                </View>
            </View>
        );
    }
}

export const styles = StyleSheet.create({
    container: {
        width: "100%",
        height: 350,
        position: "absolute",
        borderRadius: 30.5,
        backgroundColor: "#2A2E43",
        bottom: -275
    },
    insideBuildingContainer: {
        width: "100%",
        height: 425,
        position: "absolute",
        borderRadius: 30.5,
        backgroundColor: "#2A2E43",
        bottom: -275
    },
    moreDetails: {
        width: "100%",
        height: "100%",
        position: "absolute",
        backgroundColor: "#2A2E43",
        alignItems: "center",
        justifyContent: "space-between"
    },
    arrowUp: {
        color: "#ffffff",
        left: "5%",
        top: "7%",
    },
    toggle: {
        position: "absolute",
        left: "80%",
        top: "7%"
    },
    btn: {
        position: "absolute",
        left: "65%",
        top: "5.5%",
        color: "#FFFFFF"
    },
    btnGetDirection: {
        position: "absolute",
        left: "57%",
        top: "5.5%",
        color: "#FFFFFF"
    },
    btnGetDirectionPosition: {
        left: "100%"
    },
    btnleave: {
        position: "absolute",
        left: "62%",
        top: "5.5%",
        color: "#FFFFFF",
    },
    btnText: {
        color: "#FFFFFF",
        fontFamily: "encodeSansExpanded"
    },
    mainLabel: {
        position: "absolute",
        top: "5%",
        left: "12.5%",
        color: "#FFFFFF",
        fontSize: 20,
        fontFamily: "encodeSansExpanded",
        height: "9%",
        width: "45%"
    },
    shortLabel: {
        position: "absolute",
        top: "12%",
        left: "12.5%",
        color: "#80828D",
        fontSize: 16,
        fontFamily: "encodeSansExpanded"
    },
    arrowDown: {
        color: "#ffffff",
        top: "5%",
        fontSize: 54,
        position: "absolute"
    },
    changeFloor: {
        top: "12.5%",
        left: "15%",
    }
});

export { BottomMenu };