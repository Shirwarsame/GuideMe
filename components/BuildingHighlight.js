import React from "react";
import { View } from "react-native";
import { Polygon } from "react-native-maps";
import coord from "../constants/buildingCoordinates";
import { buildingData } from "../constants/buildingData";
import { connect } from "react-redux";


function mapStateToProps(buildingName) {
    return {
        selectedBuildingName: buildingName
    }
}

function mapDispatchToProps(dispatch) {
    return {
        setSelectedBuildingName: (value) => dispatch({ type: "UPDATE_SELECTED_BUILDING", payload: value }),
    }
}

/**
 * US3 - As a user, I would like to be able to identify campus buildings and
 * distinguish them from other buildings.
 * The following function colors the campus buildings.
 */

export function BuildingHighlight (props) {
    const [buildingName, setBuildingName] = React.useState("");
    var buildings = buildingData();

    return (
        <View>
            <Polygon
                id="tap_h"
                coordinates={coord.h.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["Hall Building"].name); 
                    props.setSelectedBuildingName(buildings["Hall Building"].name);
            }}
                fillColor={buildingName == buildings["Hall Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_lb"
                coordinates={coord.lb.coordinates}
                tappable={true}
                onPress={() => {
                    setBuildingName(buildings["LB Building"].name);
                    props.setSelectedBuildingName(buildings["LB Building"].name);
                } }
                fillColor={buildingName == buildings["LB Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_gm"
                coordinates={coord.gm.coordinates}
                tappable={true}
                onPress={() => {
                    setBuildingName(buildings["GM Building"].name)
                    props.setSelectedBuildingName(buildings["GM Building"].name);
                }}
                fillColor={buildingName == buildings["GM Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_ev"
                coordinates={coord.ev.coordinates}
                tappable={true}
                onPress={() => {
                    setBuildingName(buildings["EV Building"].name)
                    props.setSelectedBuildingName(buildings["EV Building"].name);
                }}
                fillColor={buildingName == buildings["EV Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_mb"
                coordinates={coord.mb.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["MB Building"].name);
                    props.setSelectedBuildingName(buildings["MB Building"].name);
                }}
                fillColor={buildingName == buildings["MB Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_sp"
                coordinates={coord.sp.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["SP Building"].name);
                    props.setSelectedBuildingName(buildings["SP Building"].name);
                }}
                fillColor={buildingName == buildings["SP Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_cj"
                coordinates={coord.cj.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["CJ Building"].name)
                    props.setSelectedBuildingName(buildings["CJ Building"].name);
            }}
            fillColor={buildingName == buildings["CJ Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_cc"
                coordinates={coord.cc.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["CC Building"].name);
                    props.setSelectedBuildingName(buildings["CC Building"].name);
            }}
            fillColor={buildingName == buildings["CC Building"].name ? "#3ACCE1": "#a282e0"}
            />

            <Polygon
                id="tap_ad"
                coordinates={coord.ad.coordinates}
                tappable={true}
                onPress={() => {
                    setBuildingName(buildings["AD Building"].name);
                    props.setSelectedBuildingName(buildings["AD Building"].name);
                }}
                fillColor={buildingName == buildings["AD Building"].name ? "#3ACCE1": "#a282e0"}
                />

            <Polygon
                id="tap_gn"
                coordinates={coord.gn.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["Grey Nuns"].name)
                    props.setSelectedBuildingName(buildings["Grey Nuns"].name);
                }}
                fillColor={buildingName == buildings["Grey Nuns"].name ? "#3ACCE1": "#a282e0"}
            />
            <Polygon
                id="tap_vl"
                coordinates={coord.vl.coordinates}
                tappable={true}
                onPress={() => { 
                    setBuildingName(buildings["VL Building"].name)
                    props.setSelectedBuildingName(buildings["VL Building"].name);
            }}
            fillColor={buildingName == buildings["VL Building"].name ? "#3ACCE1": "#a282e0"}
            />
        </View>
    );
}

export default connect(mapStateToProps, mapDispatchToProps)(BuildingHighlight);
