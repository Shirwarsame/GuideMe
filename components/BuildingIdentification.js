import React from 'react';
import { Text, View } from 'react-native';
import { styles } from '../screens/Map'
import { Marker } from 'react-native-maps';

function BuildingIdentification () {
    return (
        <View>
            <Marker
                coordinate={{ latitude: 45.497222, longitude: -73.578809 }}
                title={'H building'}>
                <Text style={styles.buildingIdentification}>H</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.496537, longitude: -73.577755 }}
                title={'LB building'}>
                <Text style={styles.buildingIdentification}>LB</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.495804, longitude: -73.578764 }}
                title={'HM building'}>
                <Text style={styles.buildingIdentification}>HM</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.495431, longitude: -73.578133 }}
                title={'EV building'}>
                <Text style={styles.buildingIdentification}>EV</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.495306, longitude: -73.579011 }}
                title={'JMSB building'}>
                <Text style={styles.buildingIdentification}>JMSB</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.457825, longitude: -73.641458 }}
                title={'SP'}>
                <Text style={styles.buildingIdentification}>SP</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.457516, longitude: -73.640327 }}
                title={'CJ'}>
                <Text style={styles.buildingIdentification}>CJ</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.458094, longitude: -73.639768 }}
                title={'AP'}>
                <Text style={styles.buildingIdentification}>AP</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.458314, longitude: -73.640409 }}
                title={'CC'}>
                <Text style={styles.buildingIdentification}>CC</Text>
            </Marker>

            <Marker
                coordinate={{ latitude: 45.493450, longitude: -73.576814 }}
                title={'GN'}>
                <Text style={styles.buildingIdentification}>GN</Text>
            </Marker>
        </View>
    );
}

export { BuildingIdentification };