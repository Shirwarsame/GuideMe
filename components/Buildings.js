import React from 'react';
import { View } from 'react-native';
import { Polygon } from 'react-native-maps';

const Buildings = () => {
    return(
        <View>
            <Polygon
                coordinates={[
                    { latitude: 45.497167, longitude: -73.579543 },
                    { latitude: 45.497713, longitude: -73.579032 },
                    { latitude: 45.497378, longitude: -73.578340 },
                    { latitude: 45.496830, longitude: -73.578853 },
                    { latitude: 45.497167, longitude: -73.579543 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
           />

            <Polygon
                coordinates={[
                    { latitude: 45.496708, longitude: -73.578525 },
                    { latitude: 45.497291, longitude: -73.578017 },
                    { latitude: 45.496896, longitude: -73.577102 },
                    { latitude: 45.496255, longitude: -73.577700 },
                    { latitude: 45.496708, longitude: -73.578525 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.495782, longitude: -73.579149 },
                    { latitude: 45.496129, longitude: -73.578806 },
                    { latitude: 45.495945, longitude: -73.578429 },
                    { latitude: 45.495616, longitude: -73.578742 },
                    { latitude: 45.495782, longitude: -73.579149 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.495574, longitude: -73.578653 },
                    { latitude: 45.495908, longitude: -73.578445 },
                    { latitude: 45.495416, longitude: -73.577551 },
                    { latitude: 45.495172, longitude: -73.577854 },
                    { latitude: 45.495574, longitude: -73.578653 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.495523, longitude: -73.579198 },
                    { latitude: 45.495358, longitude: -73.579356 },
                    { latitude: 45.495014, longitude: -73.578734 },
                    { latitude: 45.495200, longitude: -73.578530 },
                    { latitude: 45.495523, longitude: -73.579198 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.457464, longitude: -73.642074 },
                    { latitude: 45.458325, longitude: -73.641411 },
                    { latitude: 45.458165, longitude: -73.640985 },
                    { latitude: 45.457528, longitude: -73.641481 },
                    { latitude: 45.457198, longitude: -73.640661 },
                    { latitude: 45.457002, longitude: -73.640830 },
                    { latitude: 45.457464, longitude: -73.642074 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.457335, longitude: -73.640715 },
                    { latitude: 45.457179, longitude: -73.640400 },
                    { latitude: 45.457617, longitude: -73.640051 },
                    { latitude: 45.457826, longitude: -73.640483 },
                    { latitude: 45.457335, longitude: -73.640715 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.458521, longitude: -73.640679 },
                    { latitude: 45.458377, longitude: -73.640794 },
                    { latitude: 45.458076, longitude: -73.639997 },
                    { latitude: 45.458221, longitude: -73.639898 },
                    { latitude: 45.458521, longitude: -73.640679}
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />

            <Polygon
                coordinates={[
                    { latitude: 45.457901, longitude: -73.640113 },
                    { latitude: 45.458378, longitude: -73.639742 },
                    { latitude: 45.458243, longitude: -73.639460 },
                    { latitude: 45.457807, longitude: -73.639814 },
                    { latitude: 45.457901, longitude: -73.640113 }
                ]}
                fillColor="rgba(76, 79, 98, 0.7)"
            />
            
        </View>
    )
}

export { Buildings };
