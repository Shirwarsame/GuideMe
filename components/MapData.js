import { ClassRooms } from "../constants/ClassRooms";
import { buildingData } from "../constants/buildingData";

//TODO: Add more unit tests.
/**
 * This method will return an object containing all the elements necessary to display in the MoreDetails
 * screen. No need to create method inside MoreDetails, everything is taken care of here.
 * 
 * @param {*} value is value passed from the BottomMenu to be searched in the constant Building file.
 * @param {*} list is the list of buildings that we need to iterate through.
 */
const retrieveMoreDetailsData = (value, list) => {
    
    for (var key in list) {
        if (list[key].name.includes(value) || list[key].services.includes(value) || list[key].departments.includes(value)) {
            return appendInfoAndAccessibility(list[key]);
        }
    }
    return null;
}
/**
 * Because accessibility values weren't written as arrays, I need to manually create an array and then merge the 
 * result to the building object passed in parameter.
 * 
 * @param {*} building is the building object in question.
 */
const appendInfoAndAccessibility = (building) => {
    let list  = {accessibilityItems: []};
    list.accessibilityItems.push(building.hasCredit ? "Credit Card: Yes" : "Credit Card: No");
    list.accessibilityItems.push(building.hasBicycle ? "Bicycle Parking: Yes" : "Bicycle Parking: No");
    list.accessibilityItems.push(building.hasHandicap ? "Handicap Accessibility: Yes" : "Handicap Accessibility: No");
    list.accessibilityItems.push(building.hasInfocenter ? "Info-Center: Yes" : "Info-Center: No");
    list.accessibilityItems.push(building.hasParking ? "Car Parking: Yes" : "Car Parking: No");
    return ({...building, ...list});
}
/**
 * This method appends every element of an array with an index since SearchableDropDown requires 
 * an id in order to render the items.
 * @param {*} list is the list we want to append ids to.
 */
const appendIdToList = (list) => {
        
    const appendedList = list.map((element, index) => {
            return ({
                id: index + 1,
                name: element
            });
        });
    return appendedList;
}

/**
 * This method will go through each element in the buildingData and classrooms constant files and 
 * fetch the name, fullName, address, services and departements so that they can be searchable 
 * in the dropdown search list. 
 * 
 * The reason why I'm not directly calling buildingData() and ClassRooms() is because those two are lists
 * that we are going to use in two different scenario (Search and MoreDetails) and I wanted to avoid too 
 * many calls. 
 * 
 * @param {*} buildings 
 * @param {*} classrooms 
 */
export function RetrieveSearchItems (buildings, classrooms) {
    let classroomsList = [];
    let buildingNames = [];
    let buildingFullNames = [];
    let buildingAddress = [];
    let buildingServices = [];
    let buildingDepartments = [];
    for (var key in buildings) {
        buildingNames.push(buildings[key].name);
        buildingFullNames.push(buildings[key].fullName);
        buildingAddress.push(buildings[key].address);
        buildingServices.push(buildings[key].services);
        buildingDepartments.push(buildings[key].departments);
    }
    for (var key in classrooms) {
        classroomsList.push(...classrooms[key].room);
    }
    classroomsList.flat();
    buildingServices = [].concat.apply([], buildingServices);
    buildingDepartments = [].concat.apply([], buildingDepartments);
    const mergedItems = [...buildingNames, ...buildingFullNames, ...buildingAddress, ...buildingDepartments, ...buildingServices, ...classroomsList];
    return mergedItems;
}


/**
 * Will map the constant file values to the proper context : Search / MoreDetails
 * We pass the necessary values as props.
 * 
 * @param {*} props 
 */
export function MapData (props) {
    var results= [];
    let buildingList = buildingData();
    let classrooms = ClassRooms();
    switch(props.context) {
        case "More Details":
        results = retrieveMoreDetailsData(props.buildingToSearch, buildingList);
        return results;
        case "Search":
        results = appendIdToList(RetrieveSearchItems(buildingList, classrooms));
        return results
    }
    return null;
}