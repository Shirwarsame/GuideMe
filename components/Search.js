import React, { useEffect, useLayoutEffect } from "react";
import { View, StyleSheet, TouchableOpacity, AsyncStorage } from "react-native";
import SearchableDropdown from "react-native-searchable-dropdown";
import { Icon } from "react-native-elements";
import { MapData } from "./MapData";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import {getCoordinates} from "../screens/DoubleSearch";
import { store } from "../redux/reducers/index";

function mapStateToProps(state) {
  return {
      goToSearchedItem: state.goToSearchedItem
  }
}

function mapDispatchToProps(dispatch) {
  return {
    goToSearchedItemAction: (coords, name, darkMode) => dispatch({ 
      type: "UPDATE_SEARCH_BAR_VALUE_SEARCH_BAR_MARKER", 
      payload: {coordinates: coords, name: name, darkMode: darkMode}})
  }
}

/**
 * US11 - As a user, I would like to see auto-fill options when I type a query
 * US13 - As a user, I want to be able to search a building by typing its name.
 * US25 - As a user, I would like to be able to search for a room
 * US38 - As a user, I would like to search a department/faculty by name.
 * 
 * The following function renders a search bar in the map. One can search for a
 * classroom, building, department & service.
 */

function fetchData () {
  const searchInfo = MapData({context: "Search"});
  return searchInfo;
}

export function Search (props) {
  // eslint-disable-next-line no-unused-vars

  const [data, setData] = React.useState(null);
  const [searchItem, setSearchItem] = React.useState(null);
  const [isDarkedMode, setIsDarkMode] = React.useState(store.getState().isDarkMode);

  useEffect(() => {
    const items = fetchData();
    setData(items.slice(0, items.length - 3));
  }, []);

  useEffect(() => {
    updateSearchItem();
  }, [searchItem]);

  
  const goToMenu = () => {
    AsyncStorage.setItem("sideMenu", "sideMenu"); 
    props.navigation.openDrawer();
  };

  const updateSearchItem = () => {
    if (searchItem) {
      props.goToSearchedItemAction({
        latitude: searchItem.coordinates.latitude,
        longitude: searchItem.coordinates.longitude
      }, searchItem.name, isDarkedMode);
    }
  }

  useLayoutEffect(() => {
    const unsubscribe = store.subscribe(() => {
        setIsDarkMode(store.getState().isDarkMode)
    });
    return function cleanUp() {
        unsubscribe();
    }
});
  

  return (
    <View style={styles.container} testID={props.testID}>
      <View style={styles.buttonStyle}>
        <TouchableOpacity>
          <View>
            <Icon name='menu' size={30} onPress={goToMenu}></Icon>
          </View>
        </TouchableOpacity>
      </View>
      <SearchableDropdown
        onTextChange={val => val} //This must be here (does nothing)
        onItemSelect={item => {  setSearchItem({name: item.name, coordinates: getCoordinates(item.name)});  }}
        textInputStyle={styles.textInputStyle}
        itemStyle={styles.itemStyle}
        containerStyle={styles.test}
        itemTextStyle={styles.itemTextStyle}
        itemsContainerStyle={styles.itemsContainerStyle}
        items={data}
        placeholder="Where to?"
        resetValue={true}
      />
    </View>
  );
}

Search.propTypes = {
  testID: PropTypes.string,
  navigation: PropTypes.object
};

export default connect(mapStateToProps, mapDispatchToProps)(Search);


export const styles = StyleSheet.create({
  buttonStyle: {
    width: "15%",
    borderRightWidth: 0,
    backgroundColor: "#fff",
    justifyContent: "center",
    alignItems: "center",
    borderColor: "#ccc",
    borderWidth: 1,
    maxHeight: 44,
    borderBottomStartRadius: 20
  },
  test: {
    width: "90%"
  },
  container: {
    width: "80%",
    position: "absolute",
    top: 50,
    left: "10%",
    flexDirection: "row",
    flex: 1,
    justifyContent: "center",
  },
  containerStyle: {
    padding: 5
  },
  textInputStyle: {
    padding: 12,
    borderWidth: 1,
    borderLeftWidth: 0,
    borderColor: "#ccc",
    backgroundColor: "#FFFFFF",
    fontFamily: "encodeSansExpanded"
  },
  itemStyle: {
    padding: 10,
    marginTop: 2,
    backgroundColor: "#FAF9F8",
    borderColor: "#bbb",
    borderWidth: 1,
    borderRadius: 10,
  },
  itemTextStyle: {
    color: "#222",
    fontFamily: "encodeSansExpanded"
  },
  itemsContainerStyle: {
    maxHeight: "60%",
  }
});
