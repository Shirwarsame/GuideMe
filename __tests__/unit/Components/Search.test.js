import React from "react";
import { Search } from "../../../components/Search";
import renderer from "react-test-renderer";
import { RetrieveSearchItems } from "../../../components/MapData";
import { buildingData } from "../../../constants/buildingData";
import { ClassRooms } from "../../../constants/ClassRooms";
import { shallow } from "enzyme";
import { Icon } from "react-native-elements";

describe("Search component", () => {

  // let wrapper;
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, "useState");
  useStateSpy.mockImplementation((init) => [init, setState]);

  beforeEach(() => {
    wrapper = shallow(<Search/>);
  });

  const searchItems = RetrieveSearchItems(buildingData(), ClassRooms())
    test("renders correctly", () => {
        const tree = renderer.create(<Search />).toJSON();
        expect(tree).toMatchSnapshot();
      });
    test("The searchItems list contains a classroom: H925", () => {
      expect(searchItems).toContain("H925");
    })
    test("The searchItems list contain a service: CSU Day Care & Nursery", () => {
      expect(searchItems).toContain("CSU Day Care & Nursery");
    })

    test("renders side menu", () => {
      const navigation = { openDrawer: () => {
          return true;
      }};
      const wrapper = shallow(<Search navigation={navigation} />);
      wrapper.find(Icon).props().onPress();
  });
    
});