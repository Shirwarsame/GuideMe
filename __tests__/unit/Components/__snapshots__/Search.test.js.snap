// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "left": "10%",
      "position": "absolute",
      "top": 50,
      "width": "80%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderBottomStartRadius": 20,
        "borderColor": "#ccc",
        "borderRightWidth": 0,
        "borderWidth": 1,
        "justifyContent": "center",
        "maxHeight": 44,
        "width": "15%",
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View>
        <View>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    keyboardShouldPersist="always"
    style={
      Object {
        "width": "90%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Where to?"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#ccc",
          "borderLeftWidth": 0,
          "borderWidth": 1,
          "fontFamily": "encodeSansExpanded",
          "padding": 12,
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
</View>
`;
